### ISSUE TITLE
Authorized addresses

### ISSUE THREAD

**Issue by robvanmieghem** [threefoldtech/rivine#54]
To comply with KYC and AML rules, it should be possible to create chains that only allow authorized addresses to transfer coins. To make this work we need:
- [x] A (hardcoded at first) list of authorization services that are allowed to authorize new addresses.
- [x] A special transaction to authorize an address
- [x] Validate if the unlockhashes of coinoutputs are authorized addresses
- [x] Send the change of a transaction to the unlockhash from the coininputs

**Comment by LeeSmet** [comment]
Some initial thoughts: 

- We will need to keep a list of authorized addresses. Adding a bucket in the consensus set DB makes the most sense to me
- The address(es) which receive coins and blockstakes from the genesis block are regular addresses, and therefore need to be authorized: 
  1) Either we hardcode these addresses to be authorized
  2) We add the validation transactions for these addresses into the genesis block
  3) Or we accept that the addresses we send things to in the genesis block are authorized, and add them to the list of authorized addresses

  The latter might be the best option as it reduces the amount of hardcoding compared to the first option. As for the second option, we might want some time before we consider an authorization txn to be confirmed, see the next point
- Since authorizing an address is a txn, it is also susceptible to forks in the chain. So we should probably also respect a small maturity period for these kind of txns.
- Building on that, we can just let the transactionpool  get the list of authorized addresses from the consensusset. If there is no maturity period, an authorizing txn could be in the same block as a payment txn to said address, which would require extensive handling insde the txnpool. Also note that since the block size is limited, it would theoretically be possible in this scenario that an authorizing txn is not included in a block, but a payment to said address is, which would make that block invalid.
 - Depending on how we structure an authorization txn, we will need to add support for this in the explorer. Though this is lower priority imo.
 - ...

Since some initial configs will need to be decided upon when the chain is started, it might be best to also take a look at #53 while all of this is going on.

**Comment by GlenDC** [comment]
I thought we were going to use one or multiple services to provide the authorization of addresses? Is this not the case?

**Comment by robvanmieghem** [comment]
@LeeSmet valid ints you have there
- The address(es) which receive coins and blockstakes from the genesis block are regular addresses, and therefore need to be authorized:
hardcode these addresses to be authorized
-maturity period for these kind of txns: really good point, 
to be really safe, 144 blocks would be ideal but very inpractical.
since the blockstake protocol is less vulnerable to attacks from the outside, let's add a todo for now, I think every blokcreator node will have the same software and rules in the beginning, let's add a backlog issue to add some maturing(.
- and indeed, an authorized address tx needs to be in the blockchain to keep the txpool handling easier.

**Comment by robvanmieghem** [comment]
@GlenDC indeed, a list of public keys of authorization services will do for now

**Comment by LeeSmet** [comment]
Leaving this mostly here for myself

## more technical overview:

- Authorized addresses are stored in a bucket in the consensusset, with the address itself as key and the txnID where the address got validated as the value (Not sure if this is needed, we'll see later on if this value is required or not)
- Since we want a maturity delay for the authorized addresses, the notion of `delayed coin outputs` from minerfees can be followed. 
- Validation rules which need to be implemented in the consensusset and txnpool modules: for every txn/block, just check if the addresses of the coinouputs/blockstakeoutputs are in the consensusset authorized addresses bucket.
- The `TransactionBuilder` will need to be updated: Extra checks need to be added to the functions which add coinoutputs / blockstakeoutputs so ensure the addresses in them are already authorized.
- The `Wallet` must be able to return addresses which are not yet authorized for them to be passed on and become authorized later.

If the `TransactionBuilder` is further modified to sent any coin/blockstake overflow back to the original address the outputs originated from, the `Wallet` should not need any modification. 

## Sidenote: 

- The `BlockCreator` does not need to be updated: it sends the block fee to the address which is also associated with the `UBSO` it used to create the block. And since every txn in the system must adhere to the additional validation rules about the authorized addresses, we know that this address is necessarily authorized already

## Structural:

In order to support the authorized addresses (mainly in the `TransactionBuilder`, possibly in the `Wallet`):
1) The respective functions must be modified to contain some branching logic that checks if a (probably global) variable is set to support authorized addresses, and if so check if the target address is authorized.
2) A new `AuthorizedAddressXXX` struct is created which contains the original `XXX` struct as an anonymous field, thus exposing the functions on the new struct, and allowing to only reimplement the functions which require modified behaviour. 
Example: 
```go
type AuthorizedAddressTransactionBuilder struct {
  TransactionBuilder
}
// interface must be modified to also allow an error to be returned
func (txb AuthorizedAddressTransactionBuilder) AddCoinOutput(output types.CoinOutput) (index uint64,
 err error) {
// validate output.UnlockHash is in the authorized addresses bucket...
// remainder of the function is a copy of the original struct
}
...
```
3) A new `AuthorizedAddressXXX` struct is created which contains the original `XXX` struct, but not as an anonymous field. This way all functions need to be proxied to the internal struct, but we can simply add validation rules before doing so, thus not needing to copy the remainder of the function and avoiding duplicate code produced by the previous approach
Example:
```go
type AuthorizedAddressTransactionBuilder struct {
  itb TransactionBuilder
}
// interface must be modified to also allow an error to be returned
func (txb AuthorizedAddressTransactionBuilder) AddCoinOutput(output types.CoinOutput) (index uint64,
 err error) {
// validate output.UnlockHash is in the authorized addresses bucket...
...
// address is authorized, continue
return txb.itb.AddCoinOutput(output)
}
```

While I don't like modifying the existing code with checks on global variables, it doesn't produce the (small amount of) duplicate code (2), and doesn't require adding 3 lines of code for every function defined by the interface (3)

## To be determined:

- What does an `authorizing transaction` look like?
- Exactly how will the AuthorizedAddresses flag be defined? global var somewhere? variable stored inside of the modules when they are initialized?...
- ...

**Comment by LeeSmet** [comment]
## Current state:

- The transactionpool checks the output addresses and rejects unauthorized addresses
- The consensusset checks the output addresses of the transactions in a block and rejects unauthorized addresses.
- Currently, in the block validator, the authorized addresses are checked before the block timestamp. As a result, a block in the future with unauthorized addresses might be rejected even though these addresses might be valid at the time were the block would be reconsidered. I think this is expected behaviour, since ideally every node has the same consensus set, and therefore the originating node should already have rejected this block. If this is a problem, we can simply move the check down the list.
- The transaction builder respends to the first input address. This address is necessarily authorized (else it could not have an `UTXO` in the first place).
- The transaction builder does not validate the output addresses that are being added, it only ensures the respend address. Because I see no reason to check this here. It would require a change to the inteface, and the transactionpool eventually is rejected by the transactionpool anyway. IF we decide to add a check here, i'd feel it be more appropriate to add it on the api endpoint itself and validate it even before creating a transactionbuilder.

## What still needs to be done

- Define exactly how a transaction which authorized an address looks
- Add support to recognize such an address and eventually add it to the authorized addresses in the consensus (much like the delayed miner fees)
- Add unit tests


Also, right now I just added some variables in the types package which define if authorized addresses are enabled and what the initial addresses are. This should probably be changed though I feel this belongs more to #53

**Comment by robvanmieghem** [comment]
blocked by #527

**Comment by robvanmieghem** [comment]
unblocked

**Comment by robvanmieghem** [comment]
best to implement this as an extension

**Comment by GlenDC** [comment]
Will implement it as an extension indeed, shipped with rivine. will be the `authcointx` (Authorised Coin Transactions) extension. This implies that the standard transaction (`0x01`) will also be controlled by this component.

On a high level our extension's controller will have 3 transfers (for now):

1. `AuthAddressUpdate`, contains the following properties:
  - AuthAddresses: a list of addresses to auth
  - DeAuthAddresses: a list of addresses to deauth
  - Signature by the authorized condition at the time of transfer

2. `AuthConditionUpdate`, contains the following properties:
  - Condition: the new condition that is authorized from than on
  - Signature by the authorized condition at the time of transfer (using the previous condition)

3. `AuthStandardTransfer`: the `0x01` transfer, but with added validation for coin inputs/outputs (only authorized addresses are allowed to send or receive coins)

**Comment by GlenDC** [comment]
In a second version (will require an entire reset of the goldchain testnet) I would like to turn `AuthStandardTransfer` into an `AuthCoinTransfer` and let it only be used for coins. But this first requires that we work issues like #514 completely out. This can be done also in the 2nd version of the goldchain testnet.

**Comment by GlenDC** [comment]
Extension is implemented in 62995eafcbdb549480f2b87dfa6d0f19b210a879 and 111f59f1101298b7064d30a4994ac223e68c04ca.

Next steps:
- [ ] add encode/decode unit tests;
- [ ] complete documentation (together with an auth flowgraph diagram)
- [ ] integrate it in goldchain, manually test it and ensure it works

**Comment by GlenDC** [comment]
Tests and documentation is added in 16b6e480cebd45927f8756962c142df7a1bf77c6.
